File: src/utils/data_wrangling/collinearity_analysis.R
# ---- Compute statistics ----
#' Compute numeric correlation matrix
#'
#' Compute a correlation matrix for all numeric columns in the dataframe.
#'
#' @param df A data frame containing numeric columns.
#'
#' @return A correlation matrix (data frame) with pairwise correlations between
#' numeric features.
compute_num_corr_matrix <- function(df) {
    numeric_df <- df |>
        dplyr::select(tidyselect::where(is.numeric))

    correlation_matrix <- stats::cor(
        numeric_df,
        method = "spearman",
        use = "pairwise.complete.obs"
    )

    return(correlation_matrix)
}

#' Compute Cramer's V
#'
#' Compute Cramer's V for two vectors.
#'
#' @param x A categorical variable (factor or character).
#' @param y Another categorical variable (factor or character).
#'
#' @return A numeric value representing Cramer's V, which ranges from 0 to 1.
compute_cramers_v <- function(x, y) {
    contingency_table <- table(x, y)
    chi2_test <- stats::chisq.test(
        contingency_table,
        correct = FALSE
    )

    n_samples <- sum(contingency_table)
    n_rows <- nrow(contingency_table)
    n_cols <- ncol(contingency_table)

    if (n_samples == 0 || min(n_rows - 1, n_cols - 1) == 0) {
        return(0)
    }

    v_value <- sqrt(
        chi2_test$statistic / (n_samples * min(n_rows - 1, n_cols - 1))
    )

    return(as.numeric(v_value))
}

#' Compute categorical correlation matrix
#'
#' Compute a correlation matrix for all the categorical columns in the data
#' frame.
#'
#' @param df A data frame containing categorical columns.
#'
#' @return A correlation matrix (data frame) with pairwise correlations between
#' categorical features.
compute_cat_corr_matrix <- function(df) {
    cat_df <- df |>
        dplyr::select(tidyselect::where(is.factor))
    feature_names <- colnames(cat_df)
    n_features <- length(feature_names)

    # Safety check if there are not enough categorical features
    if (n_features < 2) {
        stop("The provided dataframe must contain at least 2 categorical
        columns.")
    }

    # Initialize a symmetric matrix filled with 1.0 on the diagonal
    matrix <- matrix(
        1.0,
        nrow = n_features,
        ncol = n_features,
        dimnames = list(feature_names, feature_names)
    )

    # Automate pairwise extraction using nested loops
    for (i in 1:(n_features - 1)) {
        for (j in (i + 1):n_features) {
            association_score <- compute_cramers_v(cat_df[[i]], cat_df[[j]])
            matrix[i, j] <- association_score
            matrix[j, i] <- association_score
        }
    }

    return(matrix)
}

# ---- Plots ----
#' Plot correlation matrix heat map
#'
#' Automatically filters a data frame based on dtype, computes the corresponding
#' association matrix (Pearson r or Cramer's V), and renders a lower-triangular
#' heat map.
#'
#' @param df A data frame containing clinical features.
#' @param dtype Determines whether data is categorical ("cat") or numeric ("num").
#' Default is "num".
#' @param title Plot title string. Default is NULL.
#' @param threshold Absolute value threshold to display numerical labels.
#' Default is 0.4.
#'
#' @return A ggplot2 heatmap object.
plot_corr_matrix <- function(df,
                             dtype = "num",
                             title = NULL,
                             threshold = 0.5) {
    if (dtype == "num") {
        flag <- "numeric"

        # Compute numeric matrix using your helper function
        matrix_data <- compute_num_corr_matrix(df)

        # Customize column name and scale range
        coefficient <- "r"
        limits <- c(-1, 1)

        # Customize color scheme for numeric data
        fill_scale <- ggplot2::scale_fill_gradient2(
            low = "#3182bd",
            mid = "white",
            high = "#de2d26",
            midpoint = 0,
            limits = limits,
            name = coefficient
        )
    } else if (dtype == "cat") {
        flag <- "categorical"

        # Compute categorical matrix using your helper function
        matrix_data <- compute_cat_corr_matrix(df)

        # Customize column name and scale range
        coefficient <- "V"
        limits <- c(0, 1)

        # Customize color scheme for categorical data
        fill_scale <- ggplot2::scale_fill_gradientn(
            colors = c("#fff5f5", "#fecaca", "#ef4444", "#7f1d1d"),
            limits = limits,
            name = coefficient
        )
    } else {
        stop("Invalid dtype parameter. Please specify either 'num' or 'cat'.")
    }

    # Compute the coefficient's midpoint for perfect color balancing
    midpoint <- mean(limits)

    # Mask the upper triangle and diagonal to accurately preserve the LOWER
    # triangle staircase shape requested for the final report.
    matrix_data[lower.tri(matrix_data, diag = TRUE)] <- NA

    # Save the feature names in their original structural column order
    feature_order <- colnames(matrix_data)

    # Reshape the square matrix into a clean long format data frame
    df_long <- as.data.frame(matrix_data) |>
        tibble::rownames_to_column(var = "Feature_1") |>
        tidyr::pivot_longer(
            cols = -Feature_1,
            names_to = "Feature_2",
            values_to = "Value"
        ) |>
        dplyr::filter(!is.na(Value)) |>
        dplyr::mutate(
            Feature_1 = factor(Feature_1, levels = feature_order),
            Feature_2 = factor(Feature_2, levels = rev(feature_order))
        )

    # Render the triangular heat map
    p <- ggplot2::ggplot(
        df_long,
        ggplot2::aes(x = Feature_1, y = Feature_2, fill = Value)
    ) +

        # Add the tiles
        ggplot2::geom_tile(color = "#e2e8f0", linewidth = 0.4) +

        # Inject the previously defined scale
        fill_scale +

        # Add text labels
        ggplot2::geom_text(
            data = df_long |>
                dplyr::filter(abs(Value) >= threshold),
            ggplot2::aes(
                label = sprintf("%.2f", Value),
                color = abs(Value) > threshold
            ),
            size = 3.5,
            fontface = "bold",
            show.legend = FALSE
        ) +

        # Ensures text is white on dark cells and dark-blue on light cells for
        # reading safety
        ggplot2::scale_color_manual(
            values = c("TRUE" = "white", "FALSE" = "#2c3e50")
        ) +

        # Customization
        ggplot2::labs(title = title, x = NULL, y = NULL) +
        ggplot2::theme_minimal() +
        ggplot2::theme(
            plot.title = ggplot2::element_text(
                face = "bold",
                size = 16,
                margin = ggplot2::margin(b = 10)
            ),
            axis.text.x = ggplot2::element_text(
                angle = 45,
                hjust = 1,
                face = "bold",
                color = "#34495e"
            ),
            axis.text.y = ggplot2::element_text(
                face = "bold",
                color = "#34495e"
            ),
            panel.grid = ggplot2::element_blank()
        )

    return(p)
}

#' Plot Variance Inflation Factor for each feature
#'
#' Computes the Generalized Variance Inflation Factor (GVIF) for a mixed
#' data set, safely handles invariant features within complete cases, scales it
#' back to a standard VIF equivalent, and renders a professional horizontal bar
#' chart.
#'
#' @param df A data frame containing both numeric features and categorical
#' factors.
#' @param target_var Character string specifying the dependent variable to
#' predict. Default is "AF_recurrence".
#' @param title Character string for the plot title. Default is "VIF
#' Diagnostics".
#' @param x_label Character string for the horizontal axis title. Default is
#' "VIF / GVIF^2".
#' @param y_label Character string for the vertical axis title. Defaults to NULL
#' for no label.
#'
#' @return A ggplot2 object representing the publication-ready VIF bar chart.
plot_vif <- function(df,
                     target_var = "AF_recurrence",
                     title = "VIF Diagnostics",
                     x_label = "VIF / GVIF^2",
                     y_label = NULL) {
    # 1. Clean tracking columns, drop unused factors, and force target to numeric
    temp_data <- df |>
        dplyr::mutate(
            dplyr::across(tidyselect::all_of(target_var), as.numeric)
        ) |>
        droplevels()

    # 2. Extract complete cases to replicate lm()'s internal listwise row deletion
    complete_cases_subset <- temp_data |>
        tidyr::drop_na()

    # 3. Dynamic filtration: Identify and drop invariant features within complete cases
    # Extract only factor columns first
    factor_cols <- complete_cases_subset |>
        dplyr::select(tidyselect::where(is.factor))

    # Initialize as empty vector
    single_level_features <- character(0)

    # Only attempt to pivot and summarize if there is at least ONE factor column
    if (ncol(factor_cols) > 0) {
        single_level_features <- factor_cols |>
            dplyr::summarise(
                dplyr::across(tidyselect::everything(), ~ dplyr::n_distinct(.))
            ) |>
            tidyr::pivot_longer(
                cols = tidyselect::everything(),
                names_to = "feature",
                values_to = "unique_levels"
            ) |>
            dplyr::filter(unique_levels < 2) |>
            dplyr::pull(feature)
    }

    if (length(single_level_features) > 0) {
        cat(
            "\n[VIF Diagnostics] Automatically dropping columns that become invariant within complete cases:\n",
            paste(single_level_features, collapse = ", "), "\n\n"
        )
        temp_data <- temp_data |>
            dplyr::select(-tidyselect::all_of(single_level_features))
    }

    # 4. Fit the linear regression model dynamically
    formula_string <- paste(target_var, "~ .")
    regression_model <- stats::lm(
        stats::as.formula(formula_string),
        data = temp_data
    )

    # 5. Compute VIF and process structure safely based on feature data types
    vif_raw_output <- car::vif(regression_model) |>
        as.data.frame()

    if ("GVIF^(1/(2*Df))" %in% colnames(vif_raw_output)) {
        vif_df <- vif_raw_output |>
            tibble::rownames_to_column(var = "feature") |>
            dplyr::rename(Adjusted_GVIF = `GVIF^(1/(2*Df))`) |>
            dplyr::mutate(VIF_Equivalent = Adjusted_GVIF^2) |>
            dplyr::select(feature, VIF_Equivalent)
    } else {
        vif_df <- vif_raw_output |>
            tibble::rownames_to_column(var = "feature") |>
            dplyr::rename(VIF_Equivalent = 2) |>
            dplyr::select(feature, VIF_Equivalent)
    }

    # 6. Build the professional horizontal diagnostics visualization
    p <- vif_df |>
        ggplot2::ggplot(
            ggplot2::aes(
                x = stats::reorder(feature, VIF_Equivalent),
                y = VIF_Equivalent
            )
        ) +
        ggplot2::geom_bar(
            stat = "identity",
            fill = "#3b82f6",
            alpha = 0.85,
            width = 0.7
        ) +

        # Establish clinical collinearity reference lines
        ggplot2::geom_hline(
            yintercept = 5,
            linetype = "dashed",
            color = "#ef4444",
            linewidth = 0.7
        ) +
        ggplot2::annotate(
            "text",
            x = 0.7,
            y = 5.2,
            color = "#b91c1c",
            size = 3,
            fontface = "italic",
            hjust = 0,
            label = "Severe collinearity"
        ) +

        # Horizontal orientation
        ggplot2::coord_flip(clip = "off") +
        ggplot2::scale_y_continuous(
            expand = ggplot2::expansion(mult = c(0, 0.15))
        ) +

        # Titles and design setup
        ggplot2::labs(title = title, x = y_label, y = x_label) +

        # Customization
        ggplot2::theme_minimal(base_size = 11) +
        ggplot2::theme(
            plot.title = ggplot2::element_text(
                face = "bold",
                size = 14,
                margin = ggplot2::margin(b = 4)
            ),
            plot.subtitle = ggplot2::element_text(
                color = "#64748b",
                size = 9,
                margin = ggplot2::margin(b = 15)
            ),
            axis.text.y = ggplot2::element_text(
                face = "bold",
                color = "#334155"
            ),
            axis.text.x = ggplot2::element_text(color = "#475569"),
            axis.title.x = ggplot2::element_text(
                face = "bold",
                color = "#1e293b",
                margin = ggplot2::margin(t = 10)
            ),
            panel.grid.minor = ggplot2::element_blank(),
            panel.grid.major.y = ggplot2::element_blank(),
            panel.grid.major.x = ggplot2::element_line(
                color = "#f1f5f9",
                linewidth = 0.5
            )
        )

    return(p)
}

File: src/collinearity_analysis.py
import pandas as pd
import numpy as np
import scipy.stats as stats
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm


def compute_num_corr_matrix(df):
    """
    Compute numeric correlation matrix for all numeric columns in the dataframe.
    """
    numeric_df = df.select_dtypes(include=[np.number])
    if numeric_df.empty:
        return pd.DataFrame()
    return numeric_df.corr(method="spearman")


def compute_cramers_v(x, y):
    """
    Compute Cramer's V for two categorical columns.
    """
    contingency_table = pd.crosstab(x, y)
    n_samples = contingency_table.to_numpy().sum()
    n_rows, n_cols = contingency_table.shape

    if n_samples == 0 or min(n_rows - 1, n_cols - 1) == 0:
        return 0.0

    try:
        chi2, _, _, _ = stats.chi2_contingency(contingency_table, correction=False)
    except Exception:
        return 0.0

    v_value = np.sqrt(chi2 / (n_samples * min(n_rows - 1, n_cols - 1)))
    return float(v_value)


def compute_cat_corr_matrix(df):
    """
    Compute categorical correlation matrix for all categorical columns in the dataframe.
    """
    cat_df = df.select_dtypes(include=["category", "object", "bool"])
    feature_names = list(cat_df.columns)
    n_features = len(feature_names)

    if n_features < 2:
        raise ValueError(
            "The provided dataframe must contain at least 2 categorical columns."
        )

    # Initialize a symmetric matrix filled with 1.0 on the diagonal
    corr_matrix = pd.DataFrame(1.0, index=feature_names, columns=feature_names)

    # Automate pairwise extraction using nested loops
    for i in range(n_features - 1):
        for j in range(i + 1, n_features):
            col_i = feature_names[i]
            col_j = feature_names[j]
            association_score = compute_cramers_v(cat_df[col_i], cat_df[col_j])
            corr_matrix.loc[col_i, col_j] = association_score
            corr_matrix.loc[col_j, col_i] = association_score

    return corr_matrix


def plot_corr_matrix(df, dtype="Numeric", threshold=0.5, axis_ticks_size=14):
    """
    Plot correlation matrix heatmap (lower triangular masked equivalent).
    """
    if dtype == "Numeric":
        matrix_data = compute_num_corr_matrix(df)
        if matrix_data.empty:
            return None
        vmin, vmax = -1.0, 1.0
        from matplotlib.colors import LinearSegmentedColormap

        cmap = LinearSegmentedColormap.from_list(
            "custom_num", ["#3182bd", "white", "#de2d26"]
        )
        cbar_label = "s"
    elif dtype == "Categorical":
        matrix_data = compute_cat_corr_matrix(df)
        vmin, vmax = 0.0, 1.0
        from matplotlib.colors import LinearSegmentedColormap

        cmap = LinearSegmentedColormap.from_list(
            "custom_cat", ["#fff5f5", "#fecaca", "#ef4444", "#7f1d1d"]
        )
        cbar_label = "V"
    else:
        raise ValueError(
            "Invalid dtype parameter. Please specify either 'Numeric' or 'Categorical'."
        )

    # Mask the upper triangle and keep the lower triangle visible.
    # The diagonal is hidden so the matrix looks clean and focused on pairwise
    # relationships between distinct features.
    mask = np.triu(np.ones_like(matrix_data, dtype=bool))

    plt.figure(figsize=(10, 8))

    # Custom annotations to only show values >= threshold
    annot_mask = (np.abs(matrix_data) >= threshold) & (~mask)
    annot_labels = matrix_data.map(lambda v: f"{v:.1f}" if not pd.isna(v) else "")
    annot_labels = np.where(annot_mask, annot_labels, "")

    ax = sns.heatmap(
        matrix_data,
        mask=mask,
        vmin=vmin,
        vmax=vmax,
        cmap=cmap,
        annot=annot_labels,
        fmt="",
        annot_kws={"size": 10, "weight": "bold"},
        linewidths=0.4,
        linecolor="#e2e8f0",
        square=True,
        cbar_kws={"label": cbar_label, "shrink": 0.8},
    )

    # Text color styling: white on high values, dark blue on low values
    for text in ax.texts:
        if text.get_text():
            try:
                val = float(text.get_text())
                if abs(val) > threshold:
                    text.set_color("white")
                else:
                    text.set_color("#2c3e50")
            except ValueError:
                text.set_color("#2c3e50")

    plt.title(
        f"{dtype} Correlation Matrix",
        fontsize=16,
        fontweight="bold",
        pad=15,
        loc="left",
    )
    plt.xticks(
        rotation=45,
        ha="right",
        fontsize=axis_ticks_size,
        fontweight="bold",
        color="#34495e",
    )
    plt.yticks(
        rotation=45,
        va="top",
        fontsize=axis_ticks_size,
        fontweight="bold",
        color="#34495e",
    )
    plt.tight_layout()
    return plt.gcf()


def plot_vif(
    df,
    threshold=5.0,
    target_var="AF_recurrence",
    title="VIF Diagnostics",
    x_label="VIF / GVIF^2",
    y_label=None,
):
    """
    Computes Generalized Variance Inflation Factor (GVIF) equivalent VIF for each feature,
    and plots a professional horizontal bar chart.
    """
    temp_data = df.copy()
    if target_var in temp_data.columns:
        temp_data[target_var] = pd.to_numeric(temp_data[target_var], errors="coerce")

    complete_cases_subset = temp_data.dropna()

    # Identify invariant columns (nunique < 2) in complete cases
    single_level_features = []
    for col in complete_cases_subset.columns:
        if col != target_var and complete_cases_subset[col].nunique() < 2:
            single_level_features.append(col)

    if single_level_features:
        print(
            f"\n[VIF Diagnostics] Automatically dropping columns that become invariant within complete cases:\n{', '.join(single_level_features)}\n"
        )
        temp_data = temp_data.drop(columns=single_level_features)

    # Split predictors and dummy-encode them
    if target_var in temp_data.columns:
        X_df = temp_data.drop(columns=[target_var]).dropna()
    else:
        X_df = temp_data.dropna()

    if X_df.empty:
        print("Warning: No complete cases available for VIF calculation.")
        return None

    # Compute GVIF for each original feature
    feature_groups = {}
    encoded_dfs = []

    for col in X_df.columns:
        # Check if column is categorical/object/bool
        if X_df[col].dtype.name in ["category", "object", "bool"] or not np.issubdtype(
            X_df[col].dtype, np.number
        ):
            dummies = pd.get_dummies(
                X_df[col], prefix=col, drop_first=True, dtype=float
            )
            feature_groups[col] = list(dummies.columns)
            encoded_dfs.append(dummies)
        else:
            feature_groups[col] = [col]
            encoded_dfs.append(X_df[[col]].astype(float))

    X_encoded = pd.concat(encoded_dfs, axis=1)
    corr_all = X_encoded.corr().to_numpy()
    cols_all = list(X_encoded.columns)

    vif_dict = {}
    for col, group in feature_groups.items():
        if not group:
            continue
        idx_group = [cols_all.index(c) for c in group]
        idx_other = [i for i, c in enumerate(cols_all) if c not in group]

        if not idx_other:
            vif_dict[col] = 1.0
            continue

        if len(group) == 1:
            dep = group[0]
            indep = [c for c in cols_all if c != dep]
            try:
                X_indep = sm.add_constant(X_encoded[indep])
                model = sm.OLS(X_encoded[dep], X_indep).fit()
                r2 = model.rsquared
                vif = 1.0 / (1.0 - r2) if r2 < 1.0 else 999.0
            except Exception:
                vif = 1.0
            vif_dict[col] = vif
        else:
            try:
                R_group = X_encoded[group].corr().to_numpy()
                other_cols = [cols_all[i] for i in idx_other]
                R_other = X_encoded[other_cols].corr().to_numpy()

                det_group = np.linalg.det(R_group)
                det_other = np.linalg.det(R_other)
                det_all = np.linalg.det(corr_all)

                if det_all == 0 or np.isnan(det_all):
                    # Add ridge to diagonal if singular
                    ridge = 1e-5 * np.eye(corr_all.shape[0])
                    det_all = np.linalg.det(corr_all + ridge)
                    det_group = np.linalg.det(R_group + 1e-5 * np.eye(R_group.shape[0]))
                    det_other = np.linalg.det(R_other + 1e-5 * np.eye(R_other.shape[0]))
                gvif = (det_group * det_other) / det_all if det_all != 0 else 1.0
                vif_dict[col] = float(np.sqrt(abs(gvif)))
            except Exception:
                vif_dict[col] = 1.0

    vif_series = pd.Series(vif_dict).sort_values(ascending=False)

    plt.figure(figsize=(10, max(6, len(vif_series) * 0.35)))
    sns.barplot(x=vif_series.values, y=vif_series.index, color="#2563eb")
    plt.axvline(threshold, linestyle="--", color="#e11d48", linewidth=1.2)
    plt.title(title, fontsize=13, fontweight="bold", pad=20, loc="left")
    plt.xlabel(x_label, fontsize=11, fontweight="bold", color="#1e293b", labelpad=10)
    if y_label:
        plt.ylabel(y_label, fontsize=11, fontweight="bold", color="#1e293b")
    plt.tight_layout()
    return plt.gcf()
